File: code-samples/webhooks/webhook-notification.rb
require 'ringcentral'
require 'dotenv'
# Remember to modify the path to where you saved your .env file!
Dotenv.load("./../.env")

# For the purpose of testing the code, we put the deliver address in the environment variable.
# Feel free to set the delivery address directly.
DELIVERY_ADDRESS = ENV['WEBHOOK_DELIVERY_ADDRESS'] + "/webhook"

#
# Create a Webhok notification and subscribe for instant SMS message notification
#
def subscribe_for_notification()
  begin
    eventFilters = ['/restapi/v1.0/account/~/extension/~/message-store/instant?type=SMS']
    bodyParams = {
      eventFilters: eventFilters,
      deliveryMode: {
        transportType: 'WebHook',
        address: DELIVERY_ADDRESS
      },
      expiresIn: 3600
    }
    endpoint = "/restapi/v1.0/subscription"
    resp = $platform.post(endpoint, payload: bodyParams)
    if (resp.status == 200)
      puts (resp.body)
      puts ("Subscription id: " + resp.body['id'])
      puts ("Ready to receive incoming SMS via WebHook.")
    else
      puts ("Webhook creation failed.")
      puts (resp.body)
    end
  rescue StandardError => e
    puts e
  end
end

#
# Read all created subscriptions
#
def read_subscriptions()
  begin
    endpoint = "/restapi/v1.0/subscription"
    resp = $platform.get(endpoint)
    if (resp.body['records'].length == 0)
      puts ("No subscription.")
    else
      for record in resp.body['records'] do
        puts JSON.pretty_generate(JSON.parse(record.to_json))
        delete_subscription(record['id'])
      end
    end
  rescue StandardError => e
    puts e
  end
end

#
# Delete a subscription identified by the subscription id
#
def delete_subscription(subscriptionId)
  begin
    endpoint = "/restapi/v1.0/subscription/" + subscriptionId
    resp = $platform.delete(endpoint)
    puts ("Subscription " + subscriptionId + " deleted.")
  rescue StandardError => e
    puts e
  end
end

# Instantiate the SDK and get the platform instance
$platform = RingCentral.new( ENV['RC_CLIENT_ID'], ENV['RC_CLIENT_SECRET'], ENV['RC_SERVER_URL'] )

# Authenticate a user using a personal JWT token
def login()
  begin
    $platform.authorize(jwt: ENV['RC_JWT'])
    subscribe_for_notification()
    #read_subscriptions()
  rescue StandardError => e
    puts ("Unable to authenticate to platform. Check credentials." + e.to_s)
  end
end

login()

File: code-samples/webhooks/webhook-notification.js
const RC = require('@ringcentral/sdk').SDK
const path = require('path')
// Remember to modify the path to where you saved your .env file!
require('dotenv').config({ path: path.resolve(__dirname, '../.env') })


// For the purpose of testing the code, we put the deliver address in the environment variable.
// Feel free to set the delivery address directly.
DELIVERY_ADDRESS = process.env.WEBHOOK_DELIVERY_ADDRESS + "/webhook"


// Instantiate the SDK and get the platform instance
var rcsdk = new RC({
    'server':       process.env.RC_SERVER_URL,
    'clientId':     process.env.RC_APP_CLIENT_ID,
    'clientSecret': process.env.RC_APP_CLIENT_SECRET
});
var platform = rcsdk.platform();

/* Authenticate a user using a personal JWT token */
platform.login({ 'jwt':  process.env.RC_USER_JWT })

platform.on(platform.events.loginSuccess, function(e){
    subscribe_for_notification()
    //read_subscriptions()
});

platform.on(platform.events.loginError, function(e){
    console.log("Unable to authenticate to platform. Check credentials.", e.message)
    process.exit(1)
});

/*
* Create a Webhok notification and subscribe for instant SMS message notification
*/
async function subscribe_for_notification() {
  var bodyParams = {
    eventFilters: ['/restapi/v1.0/account/~/extension/~/message-store/instant?type=SMS'],
    deliveryMode: {
      transportType: "WebHook",
      address: DELIVERY_ADDRESS
    },
    expiresIn: 3600
  }
  try {
    let endpoint = "/restapi/v1.0/subscription"
    var resp = await platform.post(endpoint, bodyParams)
    var jsonObj = await resp.json()
    console.log (`Subscription Id: ${jsonObj.id}`)
    console.log("Ready to receive incoming SMS via WebHook.")
  } catch (e) {
    console.log(e.message);
  }
}

/*
* Read all created subscriptions
*/
async function read_subscriptions(){
  try {
    let endpoint = "/restapi/v1.0/subscription"
    var resp = await platform.get(endpoint)
    var jsonObj = await resp.json()
    if (jsonObj.records.length == 0)
      console.log("No subscription.")
    else {
      for (var record of jsonObj.records){
        console.log(record)
        delete_subscription(record.id)
      }
    }
  } catch (e) {
    console.erlogror(e.message);
  }
}

/*
* Delete a subscription identified by the subscription id
*/
async function delete_subscription(subscriptionId){
  try {
    let endpoint = `/restapi/v1.0/subscription/${subscriptionId}`
    var resp = await platform.delete(endpoint)
    console.log (`Subscription ${subscriptionId} deleted.`)
  } catch (e) {
    console.log(e.message);
  }
}
